File: 0233-number-of-digit-one/0233-number-of-digit-one.java
class Solution {
    private int digitCount;                    // Total number of digits in n
    private char[] digitsArray;                // Array of digits representing n
    private Integer[][] memoization;           // DP memoization table [position][count of ones]

    /**
     * Counts the total number of digit '1' appearing in all numbers from 0 to n
     * @param n The upper bound number
     * @return Total count of digit '1' in range [0, n]
     */
    public int countDigitOne(int n) {
        // Convert number to char array for digit-by-digit processing
        digitsArray = String.valueOf(n).toCharArray();
        digitCount = digitsArray.length;
      
        // Initialize memoization table
        // First dimension: current position in the number
        // Second dimension: count of ones accumulated so far
        memoization = new Integer[digitCount][digitCount];
      
        // Start DFS from position 0, with 0 ones counted, and limit flag true
        return dfs(0, 0, true);
    }

    /**
     * Recursive function to count digit '1' using digit DP approach
     * @param position Current position/index in the number being constructed
     * @param onesCount Count of '1's accumulated in the current number being formed
     * @param isLimit Whether we're still bounded by the original number n
     * @return Total count of '1's in all valid numbers from current state
     */
    private int dfs(int position, int onesCount, boolean isLimit) {
        // Base case: reached the end of digits
        if (position >= digitCount) {
            return onesCount;
        }
      
        // Check memoization: only use cached result when not limited by upper bound
        if (!isLimit && memoization[position][onesCount] != null) {
            return memoization[position][onesCount];
        }
      
        // Determine the upper bound for current digit
        // If limited, can only go up to current digit of n; otherwise can use 0-9
        int upperBound = isLimit ? digitsArray[position] - '0' : 9;
      
        int totalCount = 0;
      
        // Try all possible digits from 0 to upperBound at current position
        for (int digit = 0; digit <= upperBound; digit++) {
            // Recursively count for next position
            // Update onesCount if current digit is 1
            // Update isLimit: remains true only if we're at limit AND chose the upperBound digit
            totalCount += dfs(
                position + 1, 
                onesCount + (digit == 1 ? 1 : 0), 
                isLimit && (digit == upperBound)
            );
        }
      
        // Store result in memoization table only when not limited
        // (limited states are specific to the input and shouldn't be cached)
        if (!isLimit) {
            memoization[position][onesCount] = totalCount;
        }
      
        return totalCount;
    }
}

File: 0233-number-of-digit-one/0233-number-of-digit-one.js
/**
 * Counts the total number of digit 1 appearing in all
 * non-negative integers from 0 to n.
 *
 * @param {number} n
 * @return {number}
 */
function countDigitOne(n) {
    // 숫자를 문자열로 변환
    const numberString = n.toString();
    const digitCount = numberString.length;

    // 메모이제이션 테이블
    const memo = Array.from(
        { length: digitCount },
        () => Array(digitCount).fill(-1)
    );

    /**
     * @param {number} position 현재 자릿수 위치
     * @param {number} onesCount 현재까지 등장한 1의 개수
     * @param {boolean} isLimit n의 범위에 제한되는지 여부
     * @return {number}
     */
    const dfs = (position, onesCount, isLimit) => {
        // 모든 자릿수를 확인한 경우
        if (position >= digitCount) {
            return onesCount;
        }

        // 제한이 없는 경우 메모이제이션 확인
        if (!isLimit && memo[position][onesCount] !== -1) {
            return memo[position][onesCount];
        }

        // 현재 위치에서 선택할 수 있는 최대 숫자
        const maxDigit = isLimit
            ? parseInt(numberString[position])
            : 9;

        let totalCount = 0;

        // 가능한 모든 숫자 탐색
        for (let digit = 0; digit <= maxDigit; digit++) {
            totalCount += dfs(
                position + 1,
                onesCount + (digit === 1 ? 1 : 0),
                isLimit && digit === maxDigit
            );
        }

        // 제한이 없는 경우 결과 저장
        if (!isLimit) {
            memo[position][onesCount] = totalCount;
        }

        return totalCount;
    };

    return dfs(0, 0, true);
}